source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,198 @@
+                                                   name_to_custom_plugin_map);
+  if (!benchmark_result_or.ok()) {
+    ENVOY_LOG_MISC(error, "Benchmark scoring error: {}: {}", benchmark_result_or.status().code(),
+                   benchmark_result_or.status().message());
+    return benchmark_result_or.status();
+  }
+  BenchmarkResult benchmark_result = benchmark_result_or.value();
+  for (const MetricEvaluation& evaluation : benchmark_result.metric_evaluations()) {
+    ENVOY_LOG_MISC(info, "Evaluation: {}", evaluation.DebugString());
+  }
+  return benchmark_result;
+}
+
+absl::StatusOr<AdaptiveLoadSessionOutput> AdaptiveLoadControllerImpl::PerformAdaptiveLoadSession(
+    nighthawk::client::NighthawkService::StubInterface* nighthawk_service_stub,
+    const AdaptiveLoadSessionSpec& input_spec) {
+  AdaptiveLoadSessionSpec spec = session_spec_proto_helper_.SetSessionSpecDefaults(input_spec);
+  absl::Status validation_status = session_spec_proto_helper_.CheckSessionSpec(spec);
+  if (!validation_status.ok()) {
+    return validation_status;
+  }
+  absl::flat_hash_map<std::string, MetricsPluginPtr> name_to_custom_metrics_plugin_map =
+      LoadMetricsPlugins(spec);
+  StepControllerPtr step_controller = LoadStepControllerPluginFromSpec(spec);
+  AdaptiveLoadSessionOutput output;
+
+  // Threshold specs are reproduced in the output proto for convenience.
+  for (const nighthawk::adaptive_load::MetricSpecWithThreshold& threshold :
+       spec.metric_thresholds()) {
+    *output.mutable_metric_thresholds()->Add() = threshold;
+  }
+
+  Envoy::MonotonicTime start_time = time_source_.monotonicTime();
+  std::string doom_reason;
+  do {
+    absl::StatusOr<BenchmarkResult> result_or = PerformAndAnalyzeNighthawkBenchmark(
+        nighthawk_service_stub, spec, name_to_custom_metrics_plugin_map, *step_controller,
+        spec.measuring_period());
+    if (!result_or.ok()) {
+      return result_or.status();
+    }
+    BenchmarkResult result = result_or.value();
+    *output.mutable_adjusting_stage_results()->Add() = result;
+    step_controller->UpdateAndRecompute(result);
+
+    const std::chrono::nanoseconds time_limit_ns(
+        Envoy::Protobuf::util::TimeUtil::DurationToNanoseconds(spec.convergence_deadline()));
+    const auto elapsed_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
+        time_source_.monotonicTime() - start_time);
+    if (elapsed_time_ns > time_limit_ns) {
+      return absl::DeadlineExceededError(absl::StrFormat(
+          "Failed to converge before deadline of %.2f seconds.", time_limit_ns.count() / 1e9));
+    }
+  } while (!step_controller->IsConverged() && !step_controller->IsDoomed(doom_reason));
+
+  if (step_controller->IsDoomed(doom_reason)) {
+    return absl::AbortedError(
+        absl::StrCat("Step controller determined that it can never converge: ", doom_reason));
+  }
+
+  absl::StatusOr<BenchmarkResult> result_or = PerformAndAnalyzeNighthawkBenchmark(
```

Comment:
Can you comment on why calling this again after complete.

Reply:
Done (this is the testing stage)

Reply:
BTW we have #539 to consider making the testing stage an optional thing, decided by the StepController, so this special call might go away.